In [1]:
%pip install pandas

# 1) Considere que você é o cientista de dados responsável pela análise do
comportamento e evolução do balanço final da empresa mês a mês, para tal, foi
disponibilizado um arquivo (metricas_financeiras_mensais) com duas colunas, uma
com a data de apuração do balanço financeiro e outra com o valor em reais do
balanço no mês de referência. Sendo assim, foram solicitadas as seguintes
informações para a tomada de decisões futuras:

# a) Quais os 5 meses com o melhor fechamento de balanço e quais os 5 meses
com o pior fechamento de balanço?

In [2]:
import pandas as pd
df = pd.read_csv('./metricas_financeiras_mensais.csv')

In [5]:
#verificar os tipos
df.dtypes

Data              object
Balanço Final    float64
dtype: object

In [7]:
# Mudar o tipo da coluna Data para datetime do Pandas
df['Data'] = pd.to_datetime(df['Data'])

In [8]:
# Ordenando pelos melhores valores
melhores_meses = df.sort_values(by='Balanço Final', ascending=False).head(5)
melhores_meses

Data  Balanço Final
63  2015-04-01      268988.61
13  2011-02-01      263324.75
25  2012-02-01      209121.28
114 2019-07-01      199522.44
126 2020-07-01      186999.54

In [10]:
# Ordenando pelos piores valores
piores_meses = df.sort_values(by='Balanço Final', ascending=True).head(5)
piores_meses

Data  Balanço Final
135 2021-04-01     -245650.86
111 2019-04-01     -244345.22
69  2015-10-01     -225372.41
119 2019-12-01     -223749.25
150 2022-07-01     -221165.42



1.   Primeira Etapa:
Vericar Os tipos das colunas.

2.   Segunda Etapa:
Identifiquei que a coluna data era uma string e não do tipo data.

3.   Terceira Etapa:
Converti a coluna Data para o tipo datetime, para evitar erros de comparação, ou até mesmo com a ordenação da tabela.

4.   Quarta Etapa:
Criei dataframes novos para o que foi pedido. Ordenando-os pelo valor da coluna Balanço Final e utilizando o .head(5) para mostrar apenas cinco valores como pedido.







# b) Qual foi o melhor ano para a empresa em termos de ganho total? Qual a
média de fechamento de balanço ano a ano? A média e a mediana se
aproximam? Em caso de negativa, explique o motivo

In [11]:
#Adicionar a coluna Ano ao Dataframe
df['Ano'] = df['Data'].dt.year

In [14]:
#Criar um dataframe contendo a soma dos balanços por ano
total_por_ano = df.groupby('Ano')['Balanço Final'].sum()
total_por_ano

Ano
2010    -20901.96
2011   -195567.53
2012    281891.48
2013   -355831.15
2014    358952.51
2015    171625.03
2016   -416496.13
2017    198857.32
2018   -150507.05
2019   -277207.50
2020    405696.86
2021    -24390.32
2022   -205341.53
2023    141671.49
2024   -236736.10
Name: Balanço Final, dtype: float64

In [29]:
#Pegando o melhor ano e o melhor Balanço
melhor_ano = total_por_ano.sort_values(ascending=False)
melhor_ano.head(1)

Ano
2020    405696.86
Name: Balanço Final, dtype: float64

In [22]:
#Outra abordagem para confirmar o valor da anterior
melhor_ano = pd.DataFrame([{'Ano': total_por_ano.idxmax(),
                            'Valor': total_por_ano.max()}])
melhor_ano

Ano      Valor
0  2020  405696.86

In [30]:
#Calculando a média por ano
media_por_ano = df.groupby('Ano')['Balanço Final'].mean()
media_por_ano

Ano
2010    -1741.830000
2011   -16297.294167
2012    23490.956667
2013   -29652.595833
2014    29912.709167
2015    14302.085833
2016   -34708.010833
2017    16571.443333
2018   -12542.254167
2019   -23100.625000
2020    33808.071667
2021    -2032.526667
2022   -17111.794167
2023    11805.957500
2024   -26304.011111
Name: Balanço Final, dtype: float64

In [31]:
#Calcular a mediana por ano
mediana_por_ano = df.groupby('Ano')['Balanço Final'].median()
mediana_por_ano

Ano
2010    -6642.165
2011   -40551.795
2012     1073.205
2013   -32070.200
2014    51555.935
2015     2128.575
2016   -34988.800
2017    15592.210
2018     1175.990
2019    -2327.590
2020    41525.580
2021   -18339.835
2022   -25965.325
2023    32148.820
2024    -3043.120
Name: Balanço Final, dtype: float64

In [39]:
#Dataframe para comparação entre Mediana e Média
comparação_media_mediana = pd.DataFrame({
    'Ano': mediana_por_ano.index,
    'Media': media_por_ano.values,
    'Mediana': mediana_por_ano.values
})
comparação_media_mediana

Ano         Media    Mediana
0   2010  -1741.830000  -6642.165
1   2011 -16297.294167 -40551.795
2   2012  23490.956667   1073.205
3   2013 -29652.595833 -32070.200
4   2014  29912.709167  51555.935
5   2015  14302.085833   2128.575
6   2016 -34708.010833 -34988.800
7   2017  16571.443333  15592.210
8   2018 -12542.254167   1175.990
9   2019 -23100.625000  -2327.590
10  2020  33808.071667  41525.580
11  2021  -2032.526667 -18339.835
12  2022 -17111.794167 -25965.325
13  2023  11805.957500  32148.820
14  2024 -26304.011111  -3043.120

# Conclusão:
Em alguns anos a média e a mediana estão mais próximas, o que significa que nesses anos os dados são mais simétricos e com poucos outliers que poderiam distorcer a média. Isso ocorre nos anos de 2017 e 2020.
Já em 2012 por outro lado, a média é muito maior que a mediana, o que pode significar que há valores elevados nesse período que distorceram o valor da média para cima.

# c) Qual o maior período de meses com o maior somatório de balanço
ininterrupto? Em que mês se inicia e finaliza esse período e qual o valor total
do mesmo? Como base para responder essas questões, considere o
seguinte exemplo:
lista_exemplo=[1,1,2,4,-5,20,-100,22]
A faixa de maior soma se dá entre o primeiro e o sexto termo da lista, visto
que, 1+1+2+4-5+20=23.

In [45]:
#Adicionando a coluna de Mês
df['Mes'] = df['Data'].dt.to_period('M')

In [46]:
#Criação de um Pandas Series com os meses e as somas dos balanços
balanco_mensal_soma = df.groupby('Mes')['Balanço Final'].sum()
balanco_mensal_soma

Mes
2010-01     32411.56
2010-02    -62423.96
2010-03    135070.95
2010-04    -94712.30
2010-05      4944.98
             ...    
2024-05    162121.17
2024-06    -91480.87
2024-07       633.56
2024-08   -158496.18
2024-09     84892.74
Freq: M, Name: Balanço Final, Length: 177, dtype: float64

In [43]:
#Algoritmo de Kadane
def kadane(soma_mensal):
    max_sum = soma_mensal.iloc[0]
    current_sum = soma_mensal.iloc[0]
    start = 0
    end = 0
    temp_start = 0

    for idx, i in enumerate(soma_mensal.iloc[1:], start=1):
        if i > current_sum + i:
            current_sum = i
            temp_start = idx
        else:
            current_sum += i

        if current_sum > max_sum:
            max_sum = current_sum
            start = temp_start
            end = idx

    return max_sum,soma_mensal.index[start],soma_mensal.index[end]

In [47]:
#Chamada e Resultado da aplicação do Algoritmo
kadane(balanco_mensal_soma)

(975952.3800000001, Period('2014-02', 'M'), Period('2015-09', 'M'))

# Conclusão:
A questão pedia por uma solução envolvendo programação dinâmica, mais precisamente utilizando o algoritmo de Kadane. Nessa questão precisei pesquisar a respeito para encontrar a solução. Prestarei os links utilizados abaixo:
https://www.youtube.com/watch?v=UncRSviH-cY
https://www.geeksforgeeks.org/largest-sum-contiguous-subarray/

O código foi feito baseado na aplicação mostrada do vídeo, apenas adaptado para a questão.

Logo o maior valor foi de 975952,38 no período de Fevereiro de 2014 até Setembro de 2015


# d) Se fosse solicitado para que você estimasse o balanço final dos meses de
outubro, novembro e dezembro de 2024 considerando o comportamento
observado nos meses anteriores, qual seria a sua estratégia para abordar o
problema?


In [49]:
#Criar a coluna Media_Movel no Pandas Series balanco_mensal_soma
balanco_mensal_soma['Media_Movel'] = balanco_mensal_soma.rolling(window=3).mean()

In [50]:
#Verificar a média dos últimos seis meses
balanco_mensal_soma['Media_Movel'].tail(6).mean()

-35442.43388888889

# Conclusão:
Utilizando a média móvel, e esperando que haja poucas variações, a estimativa média para os próximos três meses seria de aproximadamente:

**-35442.43**

# 2) Suponha que você esteja participando de um projeto voltado à mitigação de fraudes,
e para tal, você deve analisar dados diversos, tais como o CPF, Telefone Celular
(Com DDD incluso) e entre outros. Com isso em mente, responda às seguintes
questões:


# a) Quais critérios você utilizaria para indicar se o CPF é válido em termos
estruturais? Construa um algoritmo para automatizar esse processo e realize
um teste com os seguintes CPFs, indicando aqueles que são válidos ou não:
- 217.894.500-71
- 112.622.670-05
- 695.226.930-49
- 015.897.070-51
- 311.692.760-06


In [51]:
def validar_cpf(cpf):

    #Verificar se tem ponto e hífen
    if '.' and '-' not in cpf:
        return False

    # Remover a pontuação para fazer as verificações posteriores
    cpf = cpf.replace('.', '').replace('-', '')

    # Verificar se tem 11 dígitos
    if len(cpf) != 11:
        return False

    # Verificar se todos os dígitos são iguais
    if cpf == cpf[0] * len(cpf):
        return False

    # Calcular o primeiro dígito verificador
    soma = sum(int(cpf[i]) * (10 - i) for i in range(9))
    primeiro_digito = 11 - (soma % 11)
    primeiro_digito = 0 if primeiro_digito >= 10 else primeiro_digito

    # Calcular o segundo dígito verificador
    soma = sum(int(cpf[i]) * (11 - i) for i in range(10))
    segundo_digito = 11 - (soma % 11)
    segundo_digito = 0 if segundo_digito >= 10 else segundo_digito

    # Verificar se os dígitos são válidos
    return cpf[9] == str(primeiro_digito) and cpf[10] == str(segundo_digito)

In [56]:
#Lista de exemplos dada na tarefa
cpfs = ["217.894.500-71", "112.622.670-05", "695.226.930-49", "015.897.070-51",
        "311.692.760-06"]

In [57]:
#Verificação da lista de CPFs
for cpf in cpfs:
    print(f"CPF {cpf} é {'válido' if validar_cpf(cpf) else 'inválido'}")

CPF 217.894.500-71 é válido
CPF 112.622.670-05 é válido
CPF 695.226.930-49 é válido
CPF 015.897.070-51 é inválido
CPF 311.692.760-06 é inválido


# Conclusão:
Algoritmo criado baseado nas regras encontradas no seguinte site:
https://www.macoratti.net/alg_cpf.htm

# b) Quais critérios você utilizaria para indicar se um telefone celular é válido ou
não em termos estruturais? Construa um algoritmo para automatizar esse
processo e realize um teste com os seguintes telefones, indicando aqueles
que são válidos ou não:
- (64) 96743-8065
- (75) 89555-5682
- (99) 98362-0347
- (10) 98595-1389


In [58]:
def validar_telefone(telefone):
    # Remover pontuação
    telefone = telefone.replace('(', '').replace(')', '').replace(' ', '').replace('-', '')

    ddd_nao_existente = ['23', '25', '26', '29', '36', '39', '52', '56', '57',
                         '58', '59', '72', '76', '78',]

    telefone_ddd = telefone[0] + telefone[1]


    # Verificar se tem 11 dígitos
    if len(telefone) != 11:
        return False

    #Verifica se o DDD começa com 0
    if telefone[0] == '0':
        return False

    #Verifica se o segundo digito do DDD é 0
    if telefone[1] == '0':
        return False

    #Verifica se o DDD está na lista de DDD's que não existem
    if telefone_ddd in ddd_nao_existente:
        return False


    # Verificar se o primeiro dígito depois do DDD é 9
    return telefone[2] == '9'

In [59]:
telefones = ["(64) 96743-8065", "(75) 89555-5682", "(99) 98362-0347", "(10) 98595-1389" ]

In [60]:
for telefone in telefones:
    print(f"Telefone {telefone} é {'válido' if validar_telefone(telefone) else 'inválido'}")

Telefone (64) 96743-8065 é válido
Telefone (75) 89555-5682 é inválido
Telefone (99) 98362-0347 é válido
Telefone (10) 98595-1389 é inválido


# Conclusão:
Algoritmo simples para realizar verificações se o número apresentado é um celular. Se começa com o número 9. Verificando também se o DDD é válido.

# c) Erros e variações de escrita são comuns em processos de preenchimento de
dados como nomes. Com isso em mente, qual estratégia você tomaria para
que em um contexto de avaliação, um modelo retornasse True para as
seguintes comparações:
- Fernando dos Santos / Fernando Santos
- Maria da Silva Rodrigues / Maria Rodrigues
- João Roberto / joao robert


In [61]:
!pip install fuzzywuzzy

In [63]:
!pip install unidecode

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 235.5/235.5 kB 7.6 MB/s eta 0:00:00


In [67]:
import unidecode

def normalize_name(name):
    # Converter para minúsculas
    name = name.lower()

    # Remover acentuação
    name = unidecode.unidecode(name)

    # Remover preposições
    prepositions = ['da', 'de', 'dos', 'do']
    name_parts = [word for word in name.split() if word not in prepositions]

    return ' '.join(name_parts)

# Lista de exemplos
names = [
    ("Fernando dos Santos", "Fernando Santos"),
    ("Maria da Silva Rodrigues", "Maria Rodrigues"),
    ("João Roberto", "joao robert")
]



In [68]:
from fuzzywuzzy import fuzz

def is_similar(name1, name2, threshold=80):
    # Normalizar os nomes
    name1 = normalize_name(name1)
    name2 = normalize_name(name2)

    # Calcular similaridade fuzzy
    similarity = fuzz.ratio(name1, name2)

    return similarity >= threshold

# Testando com os exemplos
for name1, name2 in names:
    if is_similar(name1, name2):
        print(f"{name1} e {name2}: True")
    else:
        print(f"{name1} e {name2}: False")


Fernando dos Santos e Fernando Santos: True
Maria da Silva Rodrigues e Maria Rodrigues: True
João Roberto e joao robert: True


#Conclusão
Nessa questão meu maior problema foi encontrar similaridades entre Roberto e robert. Pesquisando pelo reddit dei de cara com essa biblioteca fuzzywuzzy que não conhecia. Mas pelos exemplos percebi que seria o que estava buscando, então apliquei ela utilizando os exemplos desse site:
https://www.geeksforgeeks.org/fuzzywuzzy-python-library/

# 3) Solicitaram para você que seja criado um modelo de apuração de métricas de
processamento em relação ao valor, quantidade de vendas e método de pagamento.
Para tal, uma base de dados (metricas_dados_de_pagamento) foi disponibilizada
como referencial, contendo as colunas user_id (Representa o parceiro analisado),
mes_ano, metodo_pagamento, status_finalizacao, qtd_total (Quantidade de
transações) e tpv (Valor total processado). Para o fechamento de receita, são
considerados apenas os pedidos finalizados em integrado, sendo assim, os
questionamentos são os seguintes:

# a) Qual parceiro (user_id) possui o maior processamento em termos de TPV em
cada um dos meses?


In [69]:
#Criando o DataFrame a partir do csv
df = pd.read_csv('./metricas_dados_de_pagamento.csv')

#Verificando se o status de finalização contém apenas dados integrados
df_integrado = df[df['status_finalizacao'] == 'integrado']

#Criando DataFrame com soma dos TPV agrupados por mes e usuário
df_por_mes_user = df_integrado.groupby(['mes_ano', 'user_id'])['tpv'].sum(
).reset_index()

#Encontrando os usuários com o TPV máximo por mês
max_tpv_por_mes = df_por_mes_user.loc[df_por_mes_user.groupby('mes_ano')['tpv'].idxmax()]

max_tpv_por_mes

mes_ano  user_id          tpv
3511   2024-07     4057  51822500.00
8982   2024-08      652  12256841.77
16863  2024-09      652   8982559.97

# Conclusão:
Os ID's de usuários com TPV máximo são o 4057 em Julho de 2024, o 652 em Agosto de 2024 e o 652, novamente, em Setembro de 2024

# b) Em qual dos meses ocorre o pico de processamento de cartão de crédito?

In [76]:
#Criando um novo DataFrame apenas para as transações de Cartão de Crédito
df_cartao = df[df['metodo_pagamento'] == 'CreditCard']

#Contabilizando o pico de processamentos de cartao agrupados por mês
df_por_mes_cartao = df_cartao.groupby(['mes_ano'])['metodo_pagamento'].count().reset_index()

df_por_mes_cartao

mes_ano  metodo_pagamento
0  2024-07             14890
1  2024-08             13974
2  2024-09             12730

In [75]:
#Dataframe contabilizando apenas transações de Cartões com finalização integrada
df_cartao = df_integrado[df_integrado['metodo_pagamento'] == 'CreditCard']

#Soma do TPV para compras no cartão
df_tpv_cartao_por_mes = df_cartao.groupby('mes_ano')['tpv'].sum().reset_index()

#Identificando os valores de TPV por mês
max_tpv_cartao_por_mes = df_tpv_cartao_por_mes.loc[df_tpv_cartao_por_mes.groupby('mes_ano')['tpv'].idxmax()]

max_tpv_cartao_por_mes


mes_ano           tpv
0  2024-07  1.926422e+08
1  2024-08  1.430182e+08
2  2024-09  1.140509e+08

#Conclusão:
Pensei que havia dois modos de encarar esta questão. A primeira pelo pico de transações com cartão de crédito realizadas. Por isso contei a quantidade de transações e não seus valores, além de contabilizar as transações que não apenas eram integradas, mas sim todas.

Na outra abordagem, fiz pelos valores das transações para descobrir quais os picos de valores e não de transações.

Fiz dessa forma, por não ter uma compreensão exata do significado de processamento, que a questão quis abordar.

# c) Parceiros supremos são aqueles que possuem média de processamento
total superior a 300 mil reais olhando para um período de 3 meses de
processamento, com isso em mente, quantos e quais são os parceiros
supremos contidos na base?


In [81]:
#Criando DataFrame da média de TPV dos usuários por mês
df_user_media = df_por_mes_user.groupby('user_id')['tpv'].mean().reset_index()

#Filtrando o DF pelos parceiros supremos
df_parceiros_supremos = df_user_media[df_user_media['tpv'] > 300000]

df_parceiros_supremos

user_id           tpv
47         50  4.834988e+05
48         51  8.581523e+05
55         58  4.143213e+05
60         63  1.331625e+06
64         67  3.008313e+05
...       ...           ...
7950     9200  3.703962e+05
7957     9207  3.156192e+05
7973     9223  4.785545e+05
7992     9242  5.122840e+05
9586    11754  7.171150e+05

[135 rows x 2 columns]

#Conclusão:
Existem 135 parceiros supremos nos últimos 3 meses. Suas identificações estão na lista acima.

# d) Analisar a estabilidade do processamento dos parceiros é um dos pontos
vitais para a manutenção da saúde processual de uma fintech, com isso em
mente, em termos percentuais e inteiros, qual foi a oscilação do total de
parceiros com processamento maior do que zero de agosto para setembro?

In [83]:
#Filtrando o DF por tpvs positivos
df_tpv_positivo = df_integrado[df_integrado['tpv'] > 0]

#Criando o DF de tpv positivo agrupado por mês
df_tpv_positivo_por_mes = df_tpv_positivo.groupby(['mes_ano', 'user_id'])[
    'tpv'
    ].sum().reset_index()

#Contabilizando a quantidade de usuários únicos por mês com tpv positivo
df_usuarios_tpv_positivo_por_mes = df_tpv_positivo_por_mes.groupby('mes_ano')[
    'user_id'
    ].nunique().reset_index()

#Contabilizando quantidade de usuários para o mês de agosto e Setembro
usuarios_agosto = df_usuarios_tpv_positivo_por_mes[
    df_usuarios_tpv_positivo_por_mes
     [
    'mes_ano'
    ] == '2024-08']['user_id'].values[0]
usuarios_setembro = df_usuarios_tpv_positivo_por_mes[
    df_usuarios_tpv_positivo_por_mes[
    'mes_ano'
    ] == '2024-09']['user_id'].values[0]

#Calculando a porcentagem de mudança entre Agosto e Setembro
mudanca_porcentagem = ((usuarios_setembro - usuarios_agosto)
/ usuarios_agosto) *100

In [84]:
(usuarios_agosto, usuarios_setembro, mudanca_porcentagem)

(7932, 6892, -13.111447302067575)

#Conclusão:
Houve uma queda de usuários de Agosto para Setembro de 13,11%. Caindo de 7932 usuários com processamentos positivos em Agosto para 6892 usuários em Setembro.